dash-mpd-cli: enable sandbox feature on Linux

al3xtjames opened the pull request:
dash-mpd-cli 0.2.29 added experimental support for sandboxing using the Landlock LSM on Linux. By default dash-mpd-cli allows [read and execute permissions for /lib, /usr, and any paths in $PATH and $LD_LIBRARY_PATH][1]. This does not work as-is on Nix as the dynamic libraries used by the runtime binaries (such as ffmpeg) do not live under /lib or /usr. Instead, patch the dash-mpd-rs crate to allow /nix in the sandbox.

[1]: https://github.com/emarsden/dash-mpd-rs/blob/7d0ee20f5686a409e1c263c55fdd7832aca0c141/src/sandbox.rs#L44-L51

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

al3xtjames commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 474608`
Commit: `8fcc0d84de80651e998a2696d5b8d9ac0ed1ec24`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>dash-mpd-cli</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

gaelj commented on pkgs/by-name/da/dash-mpd-cli/package.nix:
attribute `stdenvNoCC` is not used, might as well remove it

al3xtjames replied:
Thanks! I went with `stdenv`.

NickCao commented on pkgs/by-name/da/dash-mpd-cli/package.nix:
```diff
@@ -24,16 +24,16 @@ let
 
-  cargoHash = "sha256-YnA/LTw9xCLSnNuFDXlsGzAiTdsst2uIDewuohkkgDU=";
+  cargoHash = "sha256-aXKSO9fz4111FIiPndmEjqwogF1/NSQ9eZfFCb3rsb0=";
 
```
Or simply set `buildFeatures`

al3xtjames replied:
I ended up patching the dash-mpd crate to allow `/nix` in the sandbox.

al3xtjames committed:
dash-mpd-cli: enable sandbox feature on Linux

dash-mpd-cli 0.2.29 added experimental support for sandboxing using the
Landlock LSM on Linux. By default dash-mpd-cli allows read and execute
permissions for /lib, /usr, and any paths in $PATH and $LD_LIBRARY_PATH
[1]. This does not work as-is on Nix as the dynamic libraries used by
the runtime binaries (such as ffmpeg) do not live under /lib or /usr.
Instead, patch the dash-mpd-rs crate to allow /nix in the sandbox.

[1]: https://github.com/emarsden/dash-mpd-rs/blob/7d0ee20f5686a409e1c263c55fdd7832aca0c141/src/sandbox.rs#L44-L51

al3xtjames commented on pkgs/by-name/da/dash-mpd-cli/package.nix:
```diff
@@ -33,11 +46,30 @@ rustPlatform.buildRustPackage (finalAttrs: {
+  '';
+
+  # Use a patched version of the dash-mpd-rs crate to allow /nix in the sandbox.
+  patches = [ ./use-patched-dash-mpd-rs.patch ];
```
`cargoPatches` doesn't work since we need to patch the derivation itself (it's for patching the lockfile), but I didn't realize that the dependencies were copied to a writable directory (`$cargoDepsCopy`). I updated the derivation to patch the crate under `$cargoDepsCopy`, which is much simpler.
